contracts/test/StakeTable.t.sol

```diff
@@ -0,0 +1,145 @@
+contract StakeTable_keyRegister_Test is Test {
+    event Register(BN254.G2Point, IStakeTable.Node);
+
+    S public stakeTable;
+    ExampleToken public token;
+    address public tokenAddress;
+    address exampleTokenCreator;
+    uint256 constant INITIAL_BALANCE = 1_000;
+
+    function setUp() public {
+        exampleTokenCreator = msg.sender;
+        vm.prank(exampleTokenCreator);
+        token = new ExampleToken(INITIAL_BALANCE);
+        tokenAddress = address(token);
+        stakeTable = new S(tokenAddress);
+    }
+
+    // TODO move to some utils library?
+    // https://ethereum.stackexchange.com/a/126928
+    /// @dev Convert some bytes to an hexadecimal string
+    function iToHex(bytes memory buffer) public pure returns (string memory) {
+        // Fixed buffer size for hexadecimal conversion
+        bytes memory converted = new bytes(buffer.length * 2);
+
+        bytes memory _base = "0123456789abcdef";
+
+        for (uint256 i = 0; i < buffer.length; i++) {
+            converted[i * 2] = _base[uint8(buffer[i]) / _base.length];
+            converted[i * 2 + 1] = _base[uint8(buffer[i]) % _base.length];
+        }
+
+        return string(abi.encodePacked("0x", converted));
+    }
+
+    /// @dev Tests the key registering process
+    function testRegister() external {
+        // Generate a BLS signature and other values using rust code
+        string[] memory cmds = new string[](3);
+        cmds[0] = "diff-test";
+        cmds[1] = "gen-bls-sig";
+        cmds[2] = iToHex(abi.encode(msg.sender));
+
+        bytes memory result = vm.ffi(cmds);
+        (
+            uint256 blsSigX,
+            uint256 blsSigY,
+            uint256 blsVkx0,
+            uint256 blsVkx1,
+            uint256 blsVky0,
+            uint256 blsVky1,
+            uint256 schnorrVKx,
+            uint256 schnorrVKy,
+            address msgSenderAddress
+        ) = abi.decode(
+            result,
+            (uint256, uint256, uint256, uint256, uint256, uint256, uint256, uint256, address)
+        );
+
+        uint64 depositAmount = 10;
+        uint64 validUntilEpoch = 5;
+
+        // Prepare for the token transfer
+        vm.prank(msgSenderAddress);
+        token.approve(address(stakeTable), depositAmount);
+
+        // Note: (x,y) coordinates for each field component must be inverted.
+        BN254.G2Point memory blsVk = BN254.G2Point(blsVkx1, blsVkx0, blsVky1, blsVky0);
+        BN254.G1Point memory sig = BN254.G1Point(blsSigX, blsSigY);
+        EdOnBN254.EdOnBN254Point memory schnorrVK = EdOnBN254.EdOnBN254Point(schnorrVKx, schnorrVKy);
+
+        assertEq(msg.sender, msgSenderAddress);
+        vm.prank(msgSenderAddress);
+
+        // Failed signature verification
+        BN254.G1Point memory badSig = BN254.P1();
+        vm.expectRevert(BLSSig.BLSSigVerificationFailed.selector);
+        stakeTable.register(
+            blsVk, schnorrVK, depositAmount, IStakeTable.StakeType.Native, badSig, validUntilEpoch
+        );
+
```

Comment:
In the tests that expect some call to `revert` we make sure the right error is raised, so I think the current test captures all the logic we want. 

I am ok to write a function for each error / the happy path but do you agree the test already covers all the execution paths? In particular the right signature is passed to the `register` function when checking for some other error to be raised.

Reply:
> In the tests that expect some call to revert we make sure the right error is raised, so I think the current test captures all the logic we want.

that's wrong.

`expectRevert` capture the next line's revert, then this function is finished, no further execution is carried out. (Line +2 is not executed because the EVM is already reverted!) 

This logic is similar to how panic handler works, there's main thread wrapping around your rust binary, then when it panic, it will give control flow to the panic handler, who will optionally log some trace if in debug mode, or directly panic if in production mode. You simply cannot go back to your rust binary again, i.e. you cannot panic twice!

Reply:
I apologize, I did more tests and my understanding of forge is wrong. Sorry for the fuss.

Reply:
No problem. It was good to [split the test](a4a17a2) into several ones anyways: I had missed [one case](https://github.com/EspressoSystems/espresso-sequencer/blob/a4a17a205560c842ea67a149322cbe4123070241/contracts/test/StakeTable.t.sol#L212) where the tokens cannot be transfered.